### PR TITLE
enclave-tls: introduce opencrypt crypto wrapper

### PR DESCRIPTION
**Pull request by uudiin**
opencrypt is a crypto wrapper based on openssl.

Signed-off-by: Tianjia Zhang <tianjia.zhang@linux.alibaba.com>

### CODE REVIEW

File: enclave-tls/src/crypto_wrappers/opencrypt/cleanup.c [hunk 1]
```diff
@@ -0,0 +1,16 @@
...
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include <stdlib.h>
+#include <enclave-tls/log.h>
+#include <enclave-tls/crypto_wrapper.h>
+#include "opencrypt.h"
+
+crypto_wrapper_err_t opencrypt_cleanup(crypto_wrapper_ctx_t *ctx)
+{
```
Comment (jiazhang0) [inline enclave-tls/src/crypto_wrappers/opencrypt/cleanup.c:12]: Please refer to wolfssl instance to add the debugging codes in each method, such as `ETLS_DEBUG("ctx %p\n", ctx);`.

File: enclave-tls/src/crypto_wrappers/opencrypt/cleanup.c [hunk 2]
```diff
@@ -0,0 +1,16 @@
...
+	struct opencrypt_ctx *octx = ctx->crypto_private;
+	free(octx);
```
Comment (jiazhang0) [inline enclave-tls/src/crypto_wrappers/opencrypt/cleanup.c:14]: add a newline here

**Review by jiazhang0 — CHANGES_REQUESTED** [review]

File: enclave-tls/src/crypto_wrappers/opencrypt/gen_pubkey_hash.c [hunk 3]
```diff
@@ -0,0 +1,35 @@
...
+
+#include <enclave-tls/log.h>
+#include <enclave-tls/crypto_wrapper.h>
+#include "opencrypt.h"
+
+#define RSA_PUBKEY_3072_RAW_LEN		398
+
+crypto_wrapper_err_t opencrypt_gen_pubkey_hash(crypto_wrapper_ctx_t *ctx,
+ 			enclave_tls_cert_algo_t algo, uint8_t *hash)
+{
```
Comment (haosanzi) [inline enclave-tls/src/crypto_wrappers/opencrypt/gen_pubkey_hash.c:14]: Please add deug info `ETLS_DEBUG("ctx %p, algo %d, hash %p\n", ctx, algo, hash);`

File: enclave-tls/src/crypto_wrappers/openssl/cleanup.c [hunk 4]
```diff
@@ -0,0 +1,17 @@
...
+#include <stdlib.h>
+#include <enclave-tls/log.h>
+#include <enclave-tls/crypto_wrapper.h>
+#include "openssl.h"
+
+crypto_wrapper_err_t openssl_cleanup(crypto_wrapper_ctx_t *ctx)
+{
+	struct openssl_ctx *octx = ctx->crypto_private;
+	ETLS_DEBUG("ctx %p\n", ctx);
+	free(octx);
```
Comment (jiazhang0) [inline enclave-tls/src/crypto_wrappers/openssl/cleanup.c:15]: forget to free key object with `RSA_free(octx->key);`

File: enclave-tls/src/crypto_wrappers/openssl/gen_privkey.c [hunk 5]
```diff
@@ -0,0 +1,63 @@
+/* Copyright (c) 2020-2021 Alibaba Cloud and Intel Corporation
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include <enclave-tls/log.h>
+#include <enclave-tls/crypto_wrapper.h>
+#include "openssl.h"
+
+crypto_wrapper_err_t __secured openssl_gen_privkey(crypto_wrapper_ctx_t *ctx,
```
Comment (jiazhang0) [inline enclave-tls/src/crypto_wrappers/openssl/gen_privkey.c:10]: remove __secured keyword.

File: enclave-tls/src/crypto_wrappers/openssl/main.c [hunk 6]
```diff
@@ -0,0 +1,39 @@
...
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include <enclave-tls/crypto_wrapper.h>
+#include <enclave-tls/log.h>
+#include <enclave-tls/cert.h>
+
+crypto_wrapper_err_t openssl_init(crypto_wrapper_ctx_t *ctx);
+crypto_wrapper_err_t __secured openssl_gen_privkey(crypto_wrapper_ctx_t *ctx,
```
Comment (jiazhang0) [inline enclave-tls/src/crypto_wrappers/openssl/main.c:11]: remove __secured keyword.

File: enclave-tls/src/crypto_wrappers/openssl/gen_cert.c [hunk 7]
```diff
@@ -0,0 +1,154 @@
...
+	}
+
+	ret = -CRYPTO_WRAPPER_ERR_CERT;
+	if (!X509_sign(cert, pkey, EVP_sha256()))
+		goto err;
+
+	der = cert_info->cert_buf;
+	len = i2d_X509(cert, &der);
+	if (len < 0)
+		goto err;
```
Comment (jiazhang0) [inline enclave-tls/src/crypto_wrappers/openssl/gen_cert.c:139]: add a newline

File: enclave-tls/src/crypto_wrappers/openssl/gen_cert.c [hunk 8]
```diff
@@ -0,0 +1,154 @@
...
+	cert_info->cert_len = len;
+
+	ETLS_DEBUG("self-signing certificate generated\n");
+	ret = CRYPTO_WRAPPER_ERR_NONE;
+
+err:
+	if (ret != CRYPTO_WRAPPER_ERR_NONE) {
```
Comment (jiazhang0) [inline enclave-tls/src/crypto_wrappers/openssl/gen_cert.c:146]: remove {

File: enclave-tls/src/crypto_wrappers/openssl/gen_cert.c [hunk 9]
```diff
@@ -0,0 +1,154 @@
...
+		ETLS_DEBUG("gen_cert(), err = %d\n", ret);
+	}
```
Comment (jiazhang0) [inline enclave-tls/src/crypto_wrappers/openssl/gen_cert.c:148]: ditto

File: enclave-tls/src/crypto_wrappers/openssl/gen_cert.c [hunk 10]
```diff
@@ -0,0 +1,154 @@
...
+	if (len < 0)
+		goto err;
+	cert_info->cert_len = len;
+
+	ETLS_DEBUG("self-signing certificate generated\n");
+	ret = CRYPTO_WRAPPER_ERR_NONE;
+
+err:
+	if (ret != CRYPTO_WRAPPER_ERR_NONE) {
+		ETLS_DEBUG("gen_cert(), err = %d\n", ret);
```
Comment (jiazhang0) [inline enclave-tls/src/crypto_wrappers/openssl/gen_cert.c:147]: change to `failed to generate certificate %d`

File: enclave-tls/src/crypto_wrappers/openssl/gen_cert.c [hunk 11]
```diff
@@ -0,0 +1,154 @@
...
+	}
+	if (cert)
+		X509_free(cert);
+	if (pkey)
+		EVP_PKEY_free(pkey);
```
Comment (jiazhang0) [inline enclave-tls/src/crypto_wrappers/openssl/gen_cert.c:152]: these resources should be recorded in ctx and then free up in cleanup().

**Review by jiazhang0 — CHANGES_REQUESTED** [review]

File: enclave-tls/src/crypto_wrappers/openssl/gen_cert.c [hunk 12]
```diff
@@ -0,0 +1,158 @@
...
+	if (!EVP_PKEY_assign_RSA(pkey, octx->key))
+		goto err;
+
+	ret = -CRYPTO_WRAPPER_ERR_NO_MEM;
+	cert = X509_new();
+	if (!cert)
+		goto err;
+
+	X509_set_version(cert, 2);
+	ASN1_INTEGER_set(X509_get_serialNumber(x509), 9527);
```
Comment (jiazhang0) [inline enclave-tls/src/crypto_wrappers/openssl/gen_cert.c:55]: Shall we use a random?

**Pull request closed without merging**